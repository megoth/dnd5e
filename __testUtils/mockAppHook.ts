import mockApp from "./mockApp";

export default function mockAppHook(mock: jest.Mock, app = mockApp()) {
  mock.mockReturnValue(app);
}
