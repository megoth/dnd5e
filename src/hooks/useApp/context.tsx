import { createContext } from "react";
import { Locale } from "../../ldo/app.typings";

const AppContext = createContext<{
  currentLocale: string;
  availableLocales: Array<Locale>;
}>({
  currentLocale: "en-US",
  availableLocales: [],
});

export default AppContext;
