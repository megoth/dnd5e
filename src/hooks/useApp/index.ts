import { useContext, useEffect } from "react";
import { AppContext } from "../../contexts/app";
import { appIsLoadingLocalizations } from "../../models/app";

export default function useApp(bundles: string[] = []) {
  const { app, setBundles } = useContext(AppContext);
  useEffect(() => {
    if (!app || !appIsLoadingLocalizations(app, bundles)) {
      return;
    }
    setBundles(bundles);
  }, [app, bundles, setBundles]);
  return app;
}
