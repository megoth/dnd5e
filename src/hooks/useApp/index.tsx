import { useContext } from "react";
import AppContext from "./context";

export default function useApp() {
  return useContext(AppContext);
}
