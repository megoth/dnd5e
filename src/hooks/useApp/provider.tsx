import React, { ReactNode, useEffect, useMemo, useState } from "react";
import { LocalizationProvider, ReactLocalization } from "@fluent/react";
import { getLocale } from "../../utils/language";
import AppContext from "./context";
import useSWR from "swr";
import Loading from "../../components/loading";
import {
  AppShapeType,
  LocaleShapeType,
  ResourceBundleShapeType,
  TranslationShapeType,
  TranslationsIndexShapeType,
} from "../../ldo/app.shapeTypes";
import { SessionInfo, useLdo, useSolidAuth } from "@ldo/solid-react";
import { hash, resourceUrl } from "../../utils/url";
import { useSearchParams } from "react-router-dom";
import { FluentBundle, FluentResource } from "@fluent/bundle";
import { userIsAdmin } from "../../utils/session";
import { namedNode } from "@rdfjs/data-model";
import { useLocalStorage } from "@uidotdev/usehooks";
import { Literal } from "@rdfjs/types";

interface Props {
  children: ReactNode;
}

function getBundles(session: SessionInfo): string[] {
  return userIsAdmin(session) ? ["global", "admin"] : ["global"];
}

export default function AppProvider({ children }: Props) {
  const { session } = useSolidAuth();
  const { dataset, getResource, getSubject } = useLdo();
  const [searchParams] = useSearchParams();
  const [currentLocale, setCurrentLocale] = useLocalStorage("locale", "en-US");
  const [bundles, setBundles] = useState<string[]>(getBundles(session));

  useEffect(() => {
    setBundles(getBundles(session));
  }, [session]);

  // fetch app index
  const { data: app } = useSWR("app", async () => {
    const appUrl =
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      import.meta.env.VITE_APP_URL ||
      new URL("/data/index.ttl#dnd5e", location.origin).toString();
    const path = resourceUrl(appUrl);
    await getResource(path).readIfUnfetched();
    return getSubject(AppShapeType, appUrl);
  });

  // fetch resource bundles
  const { data: resourceBundles } = useSWR(
    () => `resourceBundles-${app["@id"]}}-${bundles.join("-")}`,
    async () =>
      (
        await Promise.all(
          app.resourceBundle.map(async (bundle) => {
            await getResource(bundle["@id"]).readIfUnfetched();
            return getSubject(ResourceBundleShapeType, bundle["@id"]);
          }),
        )
      ).filter((bundle) => bundles.indexOf(bundle.label) !== -1),
  );

  // fetch supported languages
  const { data: supportedLocales } = useSWR(
    () => `supportedLocales-${app["@id"]}}`,
    async () => {
      return Promise.all(
        app.supportLanguage.map(async (language) => {
          await getResource(language["@id"]).readIfUnfetched();
          return getSubject(LocaleShapeType, language["@id"]);
        }),
      );
    },
  );

  useEffect(() => {
    if (!supportedLocales) return;
    // TODO: FIX so that you can load translations directly
    const requestedLocale = searchParams.get("locale");
    const locale = getLocale(
      (Array.isArray(requestedLocale) ? requestedLocale[0] : requestedLocale) ||
        currentLocale,
      supportedLocales?.map((locale) => locale.language),
    );
    setCurrentLocale(locale);
  }, [searchParams.get("locale"), supportedLocales, currentLocale]);

  // fetch FAQs
  const { isLoading: faqIsLoading } = useSWR(
    () => `faqs-${resourceBundles.map((b) => b["@id"]).join("-")}}`,
    async () =>
      Promise.all(
        resourceBundles
          .flatMap((bundle) => bundle.faqIndex)
          .map(async (index) => getResource(index["@id"]).readIfUnfetched()),
      ),
  );

  // fetch translation indices
  const { data: translations } = useSWR(
    () =>
      `translationsIndices-${resourceBundles.map((b) => b["@id"]).join("-")}-${currentLocale}`,
    async () => {
      return (
        await Promise.all(
          resourceBundles
            .flatMap((bundle) => bundle.translationsIndex)
            .map(async (index) => {
              await getResource(index["@id"]).readIfUnfetched();
              return getSubject(TranslationsIndexShapeType, index["@id"]);
            }),
        )
      ).filter((index) => !index.language || index.language === currentLocale);
    },
  );

  // get translations
  const { isLoading: translationsLoading } = useSWR(
    () => `translations-${translations.map((t) => t["@id"]).join("-")}}`,
    async () =>
      Promise.all(
        translations
          .map((index) => index.translationsResource)
          .map(async (resourceUrl) => {
            return getResource(resourceUrl["@id"]).readIfUnfetched();
          }),
      ),
  );
  const localization = useMemo(() => {
    const localization = new FluentBundle(currentLocale);
    if (!resourceBundles || translationsLoading) return null;
    resourceBundles.forEach((bundle) =>
      bundle.translationsIndex
        .map((index) => index.translationsResource["@id"])
        .flatMap((resourceUrl) => {
          return dataset
            .match(null, null, null, namedNode(resourceUrl))
            .toArray()
            .filter(
              (quad) =>
                (quad.object as Literal).language ===
                currentLocale.toLowerCase(),
            )
            .map((quad) =>
              getSubject(TranslationShapeType, quad.subject.value),
            );
        })
        .forEach((translation) =>
          localization.addResource(
            new FluentResource(
              `${hash(translation["@id"])} = ${translation.definition}`,
            ),
          ),
        ),
    );
    return new ReactLocalization([localization]);
  }, [currentLocale, resourceBundles, dataset, translationsLoading]);

  if (translationsLoading || faqIsLoading || !translations || !localization) {
    return <Loading />;
  }

  return (
    <AppContext.Provider
      value={{
        app,
        currentLocale,
        availableLocales: supportedLocales?.filter(
          (locale) => locale.language !== currentLocale,
        ),
      }}
    >
      <LocalizationProvider l10n={localization}>
        {children}
      </LocalizationProvider>
    </AppContext.Provider>
  );
}
