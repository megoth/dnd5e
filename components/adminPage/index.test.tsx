import React from "react";
import { render } from "@testing-library/react";
import AdminPage from "./index";
import mockAppHook from "../../__testUtils/mockAppHook";
import useApp from "../../src/hooks/useApp";

jest.mock("../../src/hooks/useApp");
const mockedAppHook = useApp as jest.Mock;

describe("AdminPage", () => {
  it("renders", () => {
    mockAppHook(mockedAppHook);
    const { asFragment } = render(<AdminPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
