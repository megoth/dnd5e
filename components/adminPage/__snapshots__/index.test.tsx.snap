// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminPage renders 1`] = `
<DocumentFragment>
  <header>
    <a
      class="main-container main-container--content flex focus:outline-none focus:ring-2 focus:ring-red-600"
      href="/"
    >
      <picture>
        <source
          media="(min-width: 640px)"
          srcset="/icons/app/icons8-dungeons-and-dragons-128.png"
        />
        <source
          media="(min-width: 480px)"
          srcset="/icons/app/icons8-dungeons-and-dragons-96.png"
        />
        <img
          alt="[Translation for https://example.com/global/translations.ttl#appLogoDescription does not exist]"
          src="/icons/app/icons8-dungeons-and-dragons-64.png"
        />
      </picture>
      <div
        class="font-serif text-2xl sm:text-3xl md:text-4xl self-center antialiased"
      >
        [Translation for https://example.com/global/translations.ttl#appName does not exist]
      </div>
    </a>
  </header>
  <main
    class="main-container main-container--content my-1"
  >
    <div
      class="content leading-normal max-w-prose mx-auto"
      style="hyphens: auto;"
    >
      <h1>
        [Translation for https://example.com/admin/translations.ttl#adminTitle does not exist]
      </h1>
      <p>
        [Translation for https://example.com/global/translations.ttl#workInProgress does not exist]
      </p>
    </div>
  </main>
  <footer
    class="main-container main-container--content mt-8 text-center sm:text-left xs:flex xs:space-x-4 items-center justify-center sm:justify-start"
  >
    <div
      class="md:flex-grow pb-2 flex space-x-1 justify-center sm:justify-start"
    >
      <a
        class="link"
        href="/about"
      >
        [Translation for https://example.com/global/translations.ttl#aboutTitle does not exist]
      </a>
      <a
        class="link"
        href="/faq"
      >
        [Translation for https://example.com/global/translations.ttl#faqShort does not exist]
      </a>
    </div>
    <div
      class="pb-2 flex xs:space-x-2 space-y-2 xs:space-y-0 flex-col text-center xs:flex-row xs:justify-center"
    >
      <button
        class="button"
        data-testid="dark-mode-selector-button"
        type="button"
      >
        <span>
          <span
            class="text-black text-opacity-30"
          >
            ðŸ’¡
          </span>
          Â [Translation for https://example.com/global/translations.ttl#darkModeTurnOn does not exist]
        </span>
      </button>
    </div>
  </footer>
</DocumentFragment>
`;
